l1-contracts/deploy-scripts/PrepareZKChainRegistrationCalldata.s.sol

```diff
@@ -0,0 +1,286 @@
+        address baseToken;
+        bytes diamondCutData;
+        address governance;
+        address l1SharedBridgeProxy;
+        address erc20BridgeProxy;
+    }
+
+    struct ContractsBytecodes {
+        bytes beaconProxy;
+        bytes l2SharedBridgeBytecode;
+        bytes l2SharedBridgeProxyBytecode;
+    }
+
+    Config internal config;
+    ContractsBytecodes internal bytecodes;
+
+    function run() public {
+        console.log("Preparing ZK chain registration calldata");
+
+        initializeConfig();
+
+        checkBaseTokenAddress();
+
+        IGovernance.Call[] memory calls;
+        uint256 cnt = 0;
+        if (!IBridgehub(config.bridgehub).tokenIsRegistered(config.baseToken)) {
+            calls = new IGovernance.Call[](2);
+            IGovernance.Call memory baseTokenRegistrationCall = prepareRegisterBaseTokenCall();
+            calls[cnt] = baseTokenRegistrationCall;
+            cnt++;
+        } else {
+            calls = new IGovernance.Call[](1);
+        }
+
+        IGovernance.Call memory registerChainCall = prepareRegisterHyperchainCall();
+        calls[cnt] = registerChainCall;
+        cnt++;
+
+        address l2SharedBridgeProxy = computeL2BridgeAddress();
+        IGovernance.Call memory initChainCall = prepareInitializeChainGovernanceCall(l2SharedBridgeProxy);
+
+        scheduleTransparentCalldata(calls, initChainCall);
+    }
+
+    function initializeConfig() internal {
+        // Grab config from output of l1 deployment
+        string memory root = vm.projectRoot();
+        string memory path = string.concat(root, "/script-config/prepare-registration-calldata.toml");
+        string memory toml = vm.readFile(path);
+
+        // Config file must be parsed key by key, otherwise values returned
+        // are parsed alfabetically and not by key.
+        // https://book.getfoundry.sh/cheatcodes/parse-toml
+        config.bridgehub = toml.readAddress("$.deployed_addresses.bridgehub_proxy_addr");
+        config.stateTransitionProxy = toml.readAddress("$.deployed_addresses.state_transition_proxy_addr");
+        config.l1SharedBridgeProxy = toml.readAddress("$.deployed_addresses.l1_shared_bridge_proxy_addr");
+        config.erc20BridgeProxy = toml.readAddress("$.deployed_addresses.erc20_bridge_proxy_addr");
+
+        config.chainId = toml.readUint("$.chain.chain_id");
+        config.eraChainId = toml.readUint("$.chain.era_chain_id");
+        config.chainAdmin = toml.readAddress("$.chain.admin");
+        config.diamondCutData = toml.readBytes("$.chain.diamond_cut_data");
+        config.bridgehubCreateNewChainSalt = toml.readUint("$.chain.bridgehub_create_new_chain_salt");
+        config.baseToken = toml.readAddress("$.chain.base_token_addr");
+        config.governance = toml.readAddress("$.chain.governance_addr");
+
+        bytecodes.l2SharedBridgeBytecode = Utils.readHardhatBytecode(
+            "/../l2-contracts/artifacts-zk/contracts/bridge/L2SharedBridge.sol/L2SharedBridge.json"
```

Comment:
so we're reading from the locally compiled artifacts.
It does have some bad side-effects (for example now people have to compile more stuff etc) - and it is a little bit unexpected.

Could we fetch these from the Era directly? or maybe we could submit them somewhere in github?

Reply:
I'm not sure about that. I understand that someone might have invalid contracts cached and the whole thing won't work properly, but IMO it's easier to rely on people having the right bytecodes
If we publish the code somewhere - we make the script dependent on the specific version, which makes any changes complicated and adds one more thing for us to support

Reply:
longer discussion, but the very common thing (and very hard to debug) is:

* compile stuff
* try to run -- oh, it is too old code
* git pull
* try to run again -- and bum, you've used the 'old' binaries, as you forgot to recompile..

Reply:
yeah, that's the case I thought about
but I'm still not convinced that we have to overcomplicate things and try to solve it

Reply:
it is not about overcompilicating, but about having the tool as hermetic as possible, without pulling 'surprise' dependencies.

I'd suggest that we create a subdirectory here (v0.0.1) - and put those compiled .json files there.

Reply:
created an `artifacts` directory to store the jsons, now it's important to remember to update them when we update openzeppelin libs or change the L2 bridge implementation